JoodHajeer/assignment3/JoodHajeer-3.cpp

```diff
@@ -0,0 +1,234 @@
+}
+//////////////////////////////////////////////////
+//testing begins here
+bool equalVectors(vector<string> actual_vec, vector<string> expected_vec) {
+	if (actual_vec.size() != expected_vec.size())
+		return false;
+	for (int i = 0; i < actual_vec.size(); i++)
+		if (actual_vec[i] != expected_vec[i])
+			return false;
+	return true;
+}
+void testFindWords_fillDictionary(Dictionary & dictionary_) {
+	string s[4] = { "CAT","CAR", "CART", "CARD" };
+	for (int i = 0; i<4; i++)
+		dictionary_.insert(s[i]);
+}
+void testFindWords_fillCharGrid(int rows,int coloms, string letters,vector<vector<char> >& charGrid) {
+	charGrid.resize(rows, vector<char>(coloms));
+	for (int i = 0,k=0; i<rows; i++)
+		for (int j = 0; j<coloms; j++,k++)
+			charGrid[i][j] = letters[k];
+}
+void testFindWords_emptyDictionary(Dictionary & dictionary_, vector<vector<char> > & charGrid, set<string> & words) {
+	dictionary_.clear();
+	findWords(dictionary_, charGrid, words);
+	EXPECT_EQ(0, words.size());
+}
+void testFindWords_emptyCharGrid(Dictionary & dictionary_, vector<vector<char> > & charGrid, set<string> & words) {
+	charGrid.clear();
+	findWords(dictionary_, charGrid, words);
+	EXPECT_EQ(0, words.size());
+}
+void testFindWords_dictionaryWithDuplicates(Dictionary & dictionary_, vector<vector<char> > & charGrid, set<string> & words) {;
+	findWords(dictionary_, charGrid, words);
+	vector<string>  actual_vec, expected_vec;
+	set<string> ::iterator it;
+	for (it = words.begin(); it != words.end(); it++)
+		expected_vec.push_back(*it);
+	string s[2] = { "CAT","CARD" }; 
+	for (int i = 0; i<2; i++)
+		dictionary_.insert(s[i]);//add duplicates
+	findWords(dictionary_, charGrid, words);
+	for (it = words.begin(); it != words.end(); it++)
+		actual_vec.push_back(*it);
+	//EXPECT_EQ(expected_vec, actual_vec); //not working did not find out why
+	if(!equalVectors(expected_vec, actual_vec))
+		cout << "error with dictionary with duplicates case" << endl;
+}
+void testFindWords_normalCase(Dictionary & dictionary_, vector<vector<char> > & charGrid, set<string> & words) {
+	findWords(dictionary_, charGrid, words);
+	vector<string>  actual_vec;
+	set<string> ::iterator it;
+	for (it = words.begin(); it != words.end(); it++)
+		actual_vec.push_back(*it);
+	vector<string> expected_vec {"CAR","CARD","CAT"};
+	if (!equalVectors(expected_vec, actual_vec))
+		cout << "error with the normal case" << endl;
+}
+void testFindWords_noMatchWithDictionary(Dictionary & dictionary_, vector<vector<char> > & charGrid, set<string> & words) {
+	charGrid.clear();
+	testFindWords_fillCharGrid(2, 3, "WWQUFS", charGrid);
+	findWords(dictionary_, charGrid, words);
+	EXPECT_EQ(0, words.size());
+}
+void testFindWords_gridOneCell(Dictionary & dictionary_, vector<vector<char> > & charGrid, set<string> & words) {
+	charGrid.clear();
+	testFindWords_fillCharGrid(1, 1, "C", charGrid);
+	findWords(dictionary_, charGrid, words);
+	EXPECT_EQ(0, words.size());
+}
+int main() {

```

Comment:
Some comments about testing:
It's better if the functions to do each test are encapsulated, and there is no state flowing in/out of them.
On thing you can take advantage of is c++ list-initialization. For example, if you have a function:
 `void doOneTest(const vector<string> &grid, const set<string> &dictionary, ...) `
You can call it like:
`doOneTest({"AAR", "TCD"}, {"CAT", "CAR", "CART", "CARD" }, ...`

Reply:
Sure, no problem.